File: windows/viewer_pyqt/protocol.py
"""
Protocol helpers for the PyQt5 viewer v1.2.
"""
import struct
import json
import socket

MSG_JSON = 0x01
MSG_JPEG = 0x02
MSG_FILE = 0x03
MSG_AUDIO = 0x04
HEADER_SIZE = 5

def pack_json(obj):
    payload = json.dumps(obj, separators=(",", ":")).encode("utf-8")
    return struct.pack(">BI", MSG_JSON, len(payload)) + payload

def pack_file(data):
    return struct.pack(">BI", MSG_FILE, len(data)) + data

def pack_audio(data):
    return struct.pack(">BI", MSG_AUDIO, len(data)) + data

def recv_exact(sock, n):
    buf = b""
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("socket closed")
        buf += chunk
    return buf

def recv_message(sock):
    header = recv_exact(sock, HEADER_SIZE)
    msg_type, length = struct.unpack(">BI", header)
    payload = recv_exact(sock, length)
    return msg_type, payload

def send_json(sock, obj):
    sock.sendall(pack_json(obj))

def send_file(sock, data):
    sock.sendall(pack_file(data))

def send_audio(sock, data):
    sock.sendall(pack_audio(data))

File: android/remote_control_app/lib/protocol.dart
/// Wire protocol constants and helpers for the Flutter viewer v1.2.
library;

import 'dart:convert';
import 'dart:typed_data';

class Protocol {
  static const int msgJson = 0x01;
  static const int msgJpeg = 0x02;
  static const int msgFile = 0x03;
  static const int msgAudio = 0x04;
  static const int headerSize = 5;

  static Uint8List packJson(Map<String, dynamic> obj) {
    final payload = utf8.encode(jsonEncode(obj));
    final bytes = BytesBuilder()
      ..addByte(msgJson)
      ..add(_uint32Be(payload.length))
      ..add(payload);
    return bytes.toBytes();
  }

  static Uint8List packFile(Uint8List data) {
    final bytes = BytesBuilder()
      ..addByte(msgFile)
      ..add(_uint32Be(data.length))
      ..add(data);
    return bytes.toBytes();
  }

  static Uint8List _uint32Be(int value) {
    final data = ByteData(4)..setUint32(0, value, Endian.big);
    return data.buffer.asUint8List();
  }
}

class DeviceInfo {
  final String hostname;
  final String ip;
  final int port;
  final String os;
  final String version;
  final bool passwordRequired;

  DeviceInfo({
    required this.hostname,
    required this.ip,
    required this.port,
    required this.os,
    required this.version,
    required this.passwordRequired,
  });

  factory DeviceInfo.fromJson(Map<String, dynamic> json) {
    return DeviceInfo(
      hostname: json['hostname'] ?? 'unknown',
      ip: json['ip'] ?? '',
      port: json['port'] ?? 9001,
      os: json['os'] ?? '',
      version: json['version'] ?? '',
      passwordRequired: json['password_required'] ?? false,
    );
  }
}

class FileEntry {
  final String name;
  final bool isDir;
  final int size;
  final String modified;

  FileEntry({
    required this.name,
    required this.isDir,
    required this.size,
    required this.modified,
  });

  factory FileEntry.fromJson(Map<String, dynamic> json) {
    return FileEntry(
      name: json['name'] ?? '',
      isDir: json['is_dir'] ?? false,
      size: json['size'] ?? 0,
      modified: json['modified'] ?? '',
    );
  }
}
